lib/system-flake.nix

```diff
@@ -0,0 +1,233 @@
+
+  # The global extra arguments are included in both NixOS and home-manager.
+, globalSpecialArgs ? { }
+
+  # Home-manager specific extra arguments.
+, hmSpecialArgs ? { }
+
+  # NixOS specific extra arguments.
+, nixosSpecialArgs ? { }
+
+  # Evaluates to `packages.<system>.<pname> = <unstable-channel-reference>.<pname>`.
+, packagesBuilder ? (_: { })
+
+  # Shared overlays between channels, gets applied to all `channels.<name>.input`
+, sharedOverlays ? [ ]
+
+  # Default host settings.
+, hostDefaults ? { }
+, ...
+} @ args:
+
+let
+  soxin = self;
+  soxincfg = inputs.self;
+
+  inherit (nixpkgs) lib;
+  inherit (lib) mapAttrs optionalAttrs recursiveUpdate singleton;
+  inherit (builtins) removeAttrs;
+
+  otherArguments = removeAttrs args [
+    "self"
+    "inputs"
+    "hosts"
+    "withDeploy"
+    "withSops"
+    "extraGlobalModules"
+    "extraHomeManagerModules"
+    "extraNixosModules"
+    "globalSpecialArgs"
+    "hmSpecialArgs"
+    "nixosSpecialArgs"
+    "packagesBuilder"
+    "sharedOverlays"
+  ];
+
+  hosts' =
+    mapAttrs
+      (hostname: host: (recursiveUpdate
+        {
+          specialArgs = {
+            inherit soxin soxincfg home-manager;
+
+            # the mode allows us to tell at what level we are within the modules.
+            mode = "NixOS";
+          }
+          # include the global special arguments.
+          // globalSpecialArgs
+          # include the NixOS special arguments.
+          // nixosSpecialArgs;
+        }
+
+        # pass along the hosts minus the deploy key that's specific to soxin.
+        (removeAttrs host [ "deploy" ])
+      ))
+      hosts;
+
+  deploy = mapAttrs (hostname: host: host.deploy) hosts;
+
+  soxinSystemFlake = {
+    # inherit the required fields as-is
+    inherit inputs utils;
+
+    # send self as soxincfg
+    self = soxincfg;
+
+    # set the hosts
+    hosts = hosts';
+
+    # configure the channels.
+    channels.nixpkgs.input = nixpkgs;
+    channels.unstable.input = unstable;
+
+    # Overlays which are applied to all channels.
+    sharedOverlays = [
+      # Overlay imported from this flake
+      self.overlay
+      # Nix User Repository overlay
+      nur.overlay
+    ]
+    # add the sharedOverlays from soxincfg
+    ++ sharedOverlays;
+
+    # Evaluates to `packages.<system>.<pname> = <unstable-channel-reference>.<pname>`.
+    packagesBuilder = channels:
+      recursiveUpdate
+        (import ../pkgs channels)
+        (packagesBuilder channels);
+
+    # Evaluates to `devShell.<system> = "attributeValue"`
+    devShellBuilder = channels: with channels.nixpkgs;
+      let
+        # the base devShell.
+        baseShell = mkShell {
+          name = "soxincfg";
+          buildInputs = [
+            nixpkgs-fmt
+            pre-commit
+          ];
+        };
+
+        # overlay the baseShell with things that are only necessary if the
+        # user has enabled sops support.
+        sopsShell = baseShell.overrideAttrs (oa: {
```

Comment:
Shouldn't we check if sops is enabled here?

Reply:
Yea, that's correct. fixed.